### PR TITLE
Array_product

### LINKED ISSUE #24: Challenge: Array_product

### PR DESCRIPTION
**Pull request by delvor**
- [x] PR's title includes the challenge's name and language
- [x] PR has short and clear description of the challenge
- [x] PR has appropriate labels and milestones for easy identification
- [x] PR it is assigned to the owner
- [x] reviewers are assigned
- [x] the PR contributes only one focused change
- [x] the branch is up to date with `main`/`master`
- [ ] the code works when pulled and run locally
- [ ] all conflicts are resolved (if any)
- [x] It is linked to an issue in the appropriate column of the project board
- [ ] feedback is addressed (if any, and if it is appropriate feedback.)

## README Documentation

- [x] The solution is documented in `/src/README.md`
- [x] The markdown source is formatted
- [x] Spelling and grammar is correct in all text
- [x] The markdown looks correct when you preview the file
- [x] All links and images work
- [ ] The README documents the solution's behavior, strategy(ies) and implementation(s)
- [x] There are use cases in code blocks to illustrate the function's behavior

## Python Files

- [x] There is a module header
- [ ] There is a module docstring
- [x] File names are in snake_case
- [x] Test files are named `test_<module_name>.py`

## Function Docstring

- [x] Behavior description
- [ ] Parameter description
- [x] Return value description
- [x] Include any assumptions
- [x] Include 3 or more (passing!) doctests
- [x] Include 1-2 use cases (if necessary)

## Function Implementation

- [x] The solution is not is not copied from [djeada](https://github.com/djeada/Algorithms-And-Data-Structures)
- [ ] The code is auto-formatted
- [x] The code has no (reasonable) linting mistakes
- [x] Variables are named with snake_case
- [x] The function has a clear and helpful name
- [x] The file's name matches the function name
- [ ] The code follows the strategy as simply as possible
- [ ] Variable names are clear and helpful
- [ ] Comments explain the strategy (if necessary)
- [x] There are type annotations
- [x] (challenge) The code includes defensive assertions

## Unit Test Suite

- [ ] The test class has a helpful name in PascalCase
- [ ] The test class has a docstring
- Each unit test has
  - [ ] A helpful name
  - [ ] A clear docstring
  - [ ] Only one assertion
- [ ] All tests pass
- [ ] There are tests for defensive assertions
- [ ] There are tests for boundary cases
- [ ] Test suite includes black-box and glass-box tests

### CODE REVIEW

File: src/brain_teasers/python/array_product/src/array_product.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    new_arr:list[int] = list(arr)
+    size:int = len(arr)
+    for i in range(size):
+        product:int = 1
+        for j in range(size):
+            if i != j:
+                product  *= arr[j]
+        new_arr[i] = product
+
+    return new_arr
```
Comment (JumongEric) [inline src/brain_teasers/python/array_product/src/array_product.py:42]: comments are needed

Comment (delvor) [reply src/brain_teasers/python/array_product/src/array_product.py:42]: done

**Review by JumongEric — APPROVED** [review]
